includes/class-cidr-range.php

```diff
@@ -0,0 +1,139 @@
+	 * @param int    $prefix CIDR prefix. Defaults to 48 that should affect 65,793 IPs.
+	 *
+	 * @return array
+	 */
+	public static function from_ipv6( string $ip, int $prefix = self::IPV6_PREFIX ) {
+		try {
+			// Split in address and prefix length.
+			list( $address_string, $prefix_length ) = explode( '/', "$ip/$prefix" );
+
+			// Parse the address into a binary string.
+			$address_binary = inet_pton( $address_string );
+
+			// Convert the binary string to a string with hexadecimal characters.
+			$address_hexa = bin2hex( $address_binary );
+
+			// Overwriting first address string to make sure notation is optimal.
+			$address_string = inet_ntop( $address_binary );
+
+			// Calculate the number of 'flexible' bits.
+			$flexbits = 128 - $prefix_length;
+
+			// Build the hexadecimal strings of the first and last addresses.
+			$address_hexa_first = $address_hexa;
+			$address_hexa_last  = $address_hexa;
+
+			// We start at the end of the string (which is always 32 characters long).
+			$pos = 31;
+
+			while ( $flexbits > 0 ) {
+				// Get the characters at this position and convert them to an integer.
+				$char_pos_first = hexdec( substr( $address_hexa_first, $pos, 1 ) );
+				$char_pos_last  = hexdec( substr( $address_hexa_last, $pos, 1 ) );
+
+				// First address: calculate the subnet mask. min() prevents the comparison from being negative.
+				$mask = 0xf << ( min( 4, $flexbits ) );
+
+				// AND the original against its mask.
+				$address_int_first = $char_pos_first & $mask;
+
+				// Last address: OR it with (2 ^ flexbits) - 1, with flexbits limited to 4 at a time.
+				$address_int_last = $char_pos_last | ( pow( 2, min( 4, $flexbits ) ) - 1 );
+
+				// Convert them back to hexadecimal characters.
+				$char_hexa_first = dechex( $address_int_first );
+				$char_hexa_last  = dechex( $address_int_last );
+
+				// And put those character back in their strings.
+				$address_hexa_first = substr_replace( $address_hexa_first, $char_hexa_first, $pos, 1 );
+				$address_hexa_last  = substr_replace( $address_hexa_last, $char_hexa_last, $pos, 1 );
+
+				// We processed one nibble, move to previous position.
+				$flexbits -= 4;
+				--$pos;
+			}
+
+			// Convert the hexadecimal strings to a binary string and create an IPv6 address.
+			$address_string_first = inet_ntop( hex2bin( $address_hexa_first ) );
+			$address_string_last  = inet_ntop( hex2bin( $address_hexa_last ) );
+
+			return [ $address_string_first, $address_string_last ];
+		} catch ( \Exception $e ) {
+			return [];
+		}
+	}
+
+	/**
+	 * Returns the IPv4 range for a given IPv4 address and a CIDR prefix.
+	 *
+	 * @see https://stackoverflow.com/a/41673400
+	 *
+	 * @param string $ip IPv4 address.
+	 * @param int    $prefix CIDR prefix. Defaults to 20 that should affect 4,096 IPs.
+	 *
+	 * @return array
+	 */
+	public static function from_ipv4( string $ip, int $prefix = self::IPV4_PREFIX ) {
+		try {
+			list( $ip, $mask ) = explode( '/', "$ip/$prefix" );
```

Comment:
The same question also applies to this.

Reply:
Fixed on [3c0b3ac](https://github.com/Automattic/woocommerce-payments/pull/4709/commits/3c0b3ac6d3c2cf6296ae386073176cd331d4a2a0).